### PR TITLE
Enable stricter task validation for the Gradle build

### PR DESCRIPTION
**Pull request by wolfs**
So we validate the presence of `DisableCachingByDefault` for our custom tasks in the Gradle build as well.

As a follow up to #17568.

### CODE REVIEW

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/ExplodeZipAndFindJars.groovy [hunk 1]
```diff
@@ -26,12 +26,14 @@ import org.gradle.api.file.FileSystemLocation
...
 import org.gradle.api.tasks.PathSensitive
 import org.gradle.api.tasks.PathSensitivity
+import org.gradle.work.DisableCachingByDefault
 
 import java.nio.file.Files
 import java.util.zip.ZipEntry
 import java.util.zip.ZipInputStream
 
 @CompileStatic
+@DisableCachingByDefault(because = "Only filters the input artifact")
```
Comment (lptr) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/ExplodeZipAndFindJars.groovy:36]: Does it? I thought this actually extracts stuff.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/FindGradleJars.groovy [hunk 2]
```diff
@@ -30,13 +30,16 @@ import org.gradle.api.tasks.Input
...
+import org.gradle.work.DisableCachingByDefault
 
 @CompileStatic
+@DisableCachingByDefault(because = "Only filters the input artifact")
 abstract class FindGradleJars implements TransformAction<Parameters> {
 
     @CompileStatic
     interface Parameters extends TransformParameters {
         @InputFiles
+        @PathSensitive(PathSensitivity.NAME_ONLY)
```
Comment (lptr) [inline build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/FindGradleJars.groovy:42]: Do we need to mention the change around how artifact transform warnings are handled now in the release notes?

File: build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/GenerateSubprojectsInfo.kt [hunk 3]
```diff
@@ -17,8 +17,10 @@
 package gradlebuild.buildutils.tasks
 
 import org.gradle.api.tasks.TaskAction
+import org.gradle.work.DisableCachingByDefault
 
 
+@DisableCachingByDefault(because = "Only filters the input artifact")
```
Comment (lptr) [inline build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/GenerateSubprojectsInfo.kt:23]: Doesn't look like a filter to me. This is not even an artifact transform.

Comment (lptr) [reply build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/GenerateSubprojectsInfo.kt:23]: Doesn't this need an `@OutputFile`? It seems to write to a file.

File: build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateAgpVersions.kt [hunk 4]
```diff
@@ -31,6 +32,7 @@ import javax.xml.parsers.DocumentBuilderFactory
  * Fetch the latest AGP versions and write a properties file.
  * Never up-to-date, non-cacheable.
  */
+@DisableCachingByDefault(because = "Not worth caching")
```
Comment (lptr) [inline build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateAgpVersions.kt:35]: I think here the reason is more like it cannot be cached, as it changes in-place, no?

File: build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateInitPluginTemplateVersionFile.kt [hunk 5]
```diff
@@ -26,9 +26,11 @@ import org.gradle.api.tasks.Internal
 import org.gradle.api.tasks.TaskAction
 import org.gradle.internal.util.PropertiesUtils
 import org.gradle.util.VersionNumber
+import org.gradle.work.DisableCachingByDefault
 import java.util.Properties
 
 
+@DisableCachingByDefault(because = "Not worth caching")
 abstract class UpdateInitPluginTemplateVersionFile : DefaultTask() {
```
Comment (lptr) [inline build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateInitPluginTemplateVersionFile.kt:34]: Similar to `UpdateAgpVersions`.

File: build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateReleasedVersions.kt [hunk 6]
```diff
@@ -26,9 +26,11 @@ import org.gradle.api.provider.Property
 import org.gradle.api.tasks.Internal
 import org.gradle.api.tasks.TaskAction
 import org.gradle.util.GradleVersion
+import org.gradle.work.DisableCachingByDefault
 import java.io.File
 
 
+@DisableCachingByDefault(because = "Not worth caching")
 abstract class UpdateReleasedVersions : DefaultTask() {
```
Comment (lptr) [inline build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateReleasedVersions.kt:34]: Similar to `UpdateAgpVersions`.

File: build-logic/buildquality/src/main/kotlin/gradlebuild/configcachereport/tasks/VerifyDevWorkflow.kt [hunk 7]
```diff
@@ -22,8 +22,10 @@ import org.gradle.api.tasks.InputDirectory
 import org.gradle.api.tasks.PathSensitive
 import org.gradle.api.tasks.PathSensitivity
 import org.gradle.api.tasks.TaskAction
+import org.gradle.work.DisableCachingByDefault
 
 
+@DisableCachingByDefault(because = "Not worth caching")
 abstract class VerifyDevWorkflow : DefaultTask() {
```
Comment (lptr) [inline build-logic/buildquality/src/main/kotlin/gradlebuild/configcachereport/tasks/VerifyDevWorkflow.kt:29]: We could actually cache this if it produced a file output, no? Or if we had non-file output handling. So it's not cacheable because it (currently) has no cacheable output.

File: build-logic/buildquality/src/main/kotlin/gradlebuild/quickcheck/tasks/QuickCheckTask.kt [hunk 8]
```diff
@@ -24,8 +24,10 @@ import org.gradle.cache.internal.GeneratedGradleJarCache
 import org.gradle.kotlin.dsl.*
 import org.gradle.kotlin.dsl.experiments.plugins.GradleKotlinDslKtlintConventionPlugin
 import org.gradle.kotlin.dsl.support.serviceOf
+import org.gradle.work.DisableCachingByDefault
 
 
+@DisableCachingByDefault(because = "Not worth caching")
 abstract class QuickCheckTask : DefaultTask() {
```
Comment (lptr) [inline build-logic/buildquality/src/main/kotlin/gradlebuild/quickcheck/tasks/QuickCheckTask.kt:31]: This is an abstract super-class, no?

File: build-logic/kotlin-dsl/src/main/kotlin/gradlebuild/kotlindsl/compiler/tasks/CheckKotlinCompilerEmbeddableDependencies.kt [hunk 9]
```diff
@@ -22,11 +22,13 @@ import org.gradle.api.file.ConfigurableFileCollection
...
 import org.gradle.api.tasks.OutputFile
 import org.gradle.api.tasks.TaskAction
+import org.gradle.work.DisableCachingByDefault
 
 
 /**
  * Used to validate that :kotlinCompilerEmbeddable dependencies are aligned with the original dependencies.
  */
+@DisableCachingByDefault(because = "Not worth caching")
 abstract class CheckKotlinCompilerEmbeddableDependencies : DefaultTask() {
```
Comment (lptr) [inline build-logic/kotlin-dsl/src/main/kotlin/gradlebuild/kotlindsl/compiler/tasks/CheckKotlinCompilerEmbeddableDependencies.kt:32]: Also produces no cacheable output, no?

File: build-logic/performance-testing/src/main/groovy/gradlebuild/performance/generator/tasks/RemoteProject.groovy [hunk 10]
```diff
@@ -25,11 +25,13 @@ import org.gradle.api.tasks.Input
...
 import org.gradle.api.tasks.OutputDirectory
 import org.gradle.api.tasks.TaskAction
+import org.gradle.work.DisableCachingByDefault
 
 /**
  * Checkout a project template from a git repository.
  */
 @CompileStatic
+@DisableCachingByDefault(because = "Not worth caching")
 abstract class RemoteProject extends DefaultTask {
```
Comment (lptr) [inline build-logic/performance-testing/src/main/groovy/gradlebuild/performance/generator/tasks/RemoteProject.groovy:35]: Wouldn't this make sense to cache locally only, if we had the feature?

File: build-logic/performance-testing/src/main/groovy/gradlebuild/performance/tasks/RebaselinePerformanceTests.groovy [hunk 11]
```diff
@@ -22,8 +22,10 @@ import org.gradle.api.tasks.Input
 import org.gradle.api.tasks.SourceTask
 import org.gradle.api.tasks.TaskAction
 import org.gradle.api.tasks.options.Option
+import org.gradle.work.DisableCachingByDefault
 
 @CompileStatic
+@DisableCachingByDefault(because = "Not worth caching")
 abstract class RebaselinePerformanceTests extends SourceTask {
```
Comment (lptr) [inline build-logic/performance-testing/src/main/groovy/gradlebuild/performance/tasks/RebaselinePerformanceTests.groovy:29]: This also has no cacheable output, because of the in-place modification, no?

File: build-logic/profiling/src/main/kotlin/gradlebuild/buildscan/tasks/ExtractCodeQualityBuildScanData.kt [hunk 12]
```diff
@@ -66,6 +68,7 @@ abstract class AbstractExtractCodeQualityBuildScanData : DefaultTask() {
 }
 
 
+@DisableCachingByDefault(because = "Does not have cacheable outputs")
```
Comment (lptr) [inline build-logic/profiling/src/main/kotlin/gradlebuild/buildscan/tasks/ExtractCodeQualityBuildScanData.kt:71]:
```suggestion
@DisableCachingByDefault(because = "Does not produce cacheable outputs")
```

File: build-logic/profiling/src/main/kotlin/gradlebuild/buildscan/tasks/ExtractCodeQualityBuildScanData.kt [hunk 13]
```diff
@@ -82,6 +85,7 @@ abstract class ExtractCheckstyleBuildScanData : AbstractExtractCodeQualityBuildS
...
```
Comment (lptr) [inline build-logic/profiling/src/main/kotlin/gradlebuild/buildscan/tasks/ExtractCodeQualityBuildScanData.kt:88]:
```suggestion
@DisableCachingByDefault(because = "Does not produce cacheable outputs")
```

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/UnzipTransform.java [hunk 14]
```diff
@@ -41,13 +44,15 @@
  * is located in the output directory of the transform and is named after the zipped file name
  * minus the extension.
  */
-public interface UnzipTransform extends TransformAction<TransformParameters.None> {
+@DisableCachingByDefault(because = "Not worth caching")
+public abstract class UnzipTransform implements TransformAction<TransformParameters.None> {
```
Comment (lptr) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/UnzipTransform.java:48]: Should we add a warning if an interface has a default implementation for `@TaskAction` or for `TransformAction.transform()`? Maybe create an issue to do this later?

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/FindGradleSources.java [hunk 15]
```diff
@@ -30,9 +30,9 @@
 import java.io.File;
 import java.util.Arrays;
 
-@DisableCachingByDefault(because = "Only selects locations in the input artifact")
+@DisableCachingByDefault(because = "Only filters the input artifact")
 public abstract class FindGradleSources implements TransformAction<TransformParameters.None> {
-    @PathSensitive(PathSensitivity.NONE)
+    @PathSensitive(PathSensitivity.RELATIVE)
```
Comment (lptr) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/FindGradleSources.java:35]: Why this change?

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt [hunk 16]
```diff
@@ -36,6 +36,7 @@ import java.io.File
  */
 @DisableCachingByDefault(because = "Only filters the input artifact")
 abstract class FindGradleSources : TransformAction<TransformParameters.None> {
+    @get:PathSensitive(PathSensitivity.RELATIVE)
```
Comment (lptr) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt:39]: Can we reuse the other `FindGradleSources` here somehow?

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/internal/ValidateAction.java [hunk 17]
```diff
@@ -131,7 +131,7 @@ private static DefaultTypeValidationContext createValidationContext(Class<?> top
     }
 
     private static void validateCacheabilityAnnotationPresent(Class<?> topLevelBean, boolean cacheable, Class<? extends Annotation> cacheableAnnotationClass, DefaultTypeValidationContext validationContext) {
-        if (!cacheable && topLevelBean.getAnnotation(DisableCachingByDefault.class) == null) {
+        if (!topLevelBean.isInterface() && !cacheable && topLevelBean.getAnnotation(DisableCachingByDefault.class) == null) {
```
Comment (lptr) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/internal/ValidateAction.java:134]: I think it would make sense to do this in a separate `if`, something like:

```java
if (topLevelBean.isInterface()) {
    // Won't validate interfaces
    return;
}
```

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/FindGradleJars.groovy (hunk 2 above)
Comment (wolfs) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/FindGradleJars.groovy:42]: Good question. I can add something to the upgrading doc.

File: build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/GenerateSubprojectsInfo.kt (hunk 3 above)
Comment (wolfs) [reply build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/GenerateSubprojectsInfo.kt:23]: Yeah, this should be `Not worth caching` and nothing about the filtering. The base class should probably declare the output file.

File: build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateAgpVersions.kt (hunk 4 above)
Comment (wolfs) [reply build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateAgpVersions.kt:35]: It loads data from the internet and then writes the file. This is not about changing things in-place.

File: build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateReleasedVersions.kt (hunk 6 above)
Comment (wolfs) [reply build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/UpdateReleasedVersions.kt:34]: This is different, since the current contents of the released versions file is also an input and this doesn't read data from the internet.

File: build-logic/buildquality/src/main/kotlin/gradlebuild/configcachereport/tasks/VerifyDevWorkflow.kt (hunk 7 above)
Comment (wolfs) [reply build-logic/buildquality/src/main/kotlin/gradlebuild/configcachereport/tasks/VerifyDevWorkflow.kt:29]: It doesn't have an output, though even if it would, it is not worth caching.

File: build-logic/buildquality/src/main/kotlin/gradlebuild/quickcheck/tasks/QuickCheckTask.kt (hunk 8 above)
Comment (wolfs) [reply build-logic/buildquality/src/main/kotlin/gradlebuild/quickcheck/tasks/QuickCheckTask.kt:31]: This is not an abstract super class.

File: build-logic/kotlin-dsl/src/main/kotlin/gradlebuild/kotlindsl/compiler/tasks/CheckKotlinCompilerEmbeddableDependencies.kt (hunk 9 above)
Comment (wolfs) [reply build-logic/kotlin-dsl/src/main/kotlin/gradlebuild/kotlindsl/compiler/tasks/CheckKotlinCompilerEmbeddableDependencies.kt:32]: It produces the receipt file as output.

File: build-logic/performance-testing/src/main/groovy/gradlebuild/performance/generator/tasks/RemoteProject.groovy (hunk 10 above)
Comment (wolfs) [reply build-logic/performance-testing/src/main/groovy/gradlebuild/performance/generator/tasks/RemoteProject.groovy:35]: Maybe. I suppose up-to-date checks would be good enough here.

File: build-logic/performance-testing/src/main/groovy/gradlebuild/performance/tasks/RebaselinePerformanceTests.groovy (hunk 11 above)
Comment (wolfs) [reply build-logic/performance-testing/src/main/groovy/gradlebuild/performance/tasks/RebaselinePerformanceTests.groovy:29]: Yes.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/FindGradleSources.java (hunk 15 above)
Comment (wolfs) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/FindGradleSources.java:35]: None seems to be wrong.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt (hunk 16 above)
Comment (wolfs) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt:39]: I really don't know what the goal is for these two transforms called `FindGradleSources`, and why we have two different ones. Maybe @eskatos or @bamboo have some idea what is going on here.

If we want to unify those, I think this should go in a separate PR.

File: build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/GenerateSubprojectsInfo.kt (hunk 3 above)
Comment (wolfs) [reply build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/GenerateSubprojectsInfo.kt]: Fixed.

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/internal/ValidateAction.java (hunk 17 above)
Comment (wolfs) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/internal/ValidateAction.java]: Done.

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/ExplodeZipAndFindJars.groovy (hunk 1 above)
Comment (wolfs) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/ExplodeZipAndFindJars.groovy]: You are right. Changed to `Not worth caching`.

**Comment by blindpirate** [comment]
@bot-gradle test and merge

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/UnzipTransform.java (hunk 14 above)
Comment (wolfs) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/UnzipTransform.java:48]: I created https://github.com/gradle/gradle/issues/17667.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt (hunk 16 above)
Comment (eskatos) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt:39]: I don't know/recall why we have two.
The Kotlin DSL one predates the IDE one that was introduced in https://github.com/gradle/gradle/pull/11785

Comment (wolfs) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt:39]: @rieske Do you know why we have two of those transforms?

File: build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/FindGradleJars.groovy (hunk 2 above)
Comment (wolfs) [reply build-logic/binary-compatibility/src/main/groovy/gradlebuild/binarycompatibility/transforms/FindGradleJars.groovy:42]: I tried adding something to the upgrading doc. While writing it I came to the conclusion that it is probably better not to include it. I'd treat this as a new validation which we added with `enableStricterValidation`. We don't consider that a possibly breaking change, right?

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt (hunk 16 above)
Comment (rieske) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/resolver/FindGradleSources.kt:39]: Most probably because I overlooked removing the one I added when I deleted those changes: https://github.com/gradle/gradle/pull/17683
